core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,303 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, State}
+import cats.implicits._
+import rankn._
+
+import Identifier.Constructor
+
+sealed abstract class NormalExpression {
+  val maxLambdaVar: Option[Int]
```

Comment:
it's useful to have a `maxLambda` to know when you have unbound lambdas inside your expression. e.g. when you're defining an expression inside of a lambda definition

Reply:
Can you add a comment to the meaning of None in this case? Also nit: can we make this a `def`. Note in scala you can override a `def` with a `val`. Unititalized vals can cause confusing errors, so I really only use abstract vals if I absolutely must (i.e. I am referencing a path dependent type of the val, and that can't be done on a `def`).

Reply:
`None` is for when a `NormalExpression` does not contain any expressions. eg with a `LambdaVar` and a `Literal`. All other numbers actually contain meaning that isn't communicated by this. `None` is really `-Infinity`